File: server/models/User.model.js
import mongoose from "mongoose";

const userSchema = new mongoose.Schema(
  {
    // Core business owner fields (from spec)
    businessName: {
      type: String,
      trim: true,
      default: "",
    },
    ownerName: {
      type: String,
      trim: true,
      default: "",
    },
    phone: {
      type: String,
      required: true,
      unique: true,
      trim: true,
    },
    email: {
      type: String,
      sparse: true,
      lowercase: true,
      trim: true,
    },
    address: {
      type: String,
      trim: true,
    },
    city: {
      type: String,
      trim: true,
    },
    state: {
      type: String,
      default: "Maharashtra",
      trim: true,
    },
    logoUrl: {
      type: String,
    },
    /** Shown on customer app for UPI payments (e.g. name@paytm). */
    upiId: {
      type: String,
      trim: true,
      default: "",
    },
    fcmToken: {
      type: String,
      default: null,
    },
    appVersion: {
      type: String,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    invoiceCounter: {
      type: Number,
      default: 0,
    },
    settings: {
      timezone: {
        type: String,
        default: "Asia/Kolkata",
      },
      currency: {
        type: String,
        default: "INR",
      },
      orderProcessCutoff: {
        type: String,
        default: "10:00",
      },
      autoGenerateInvoice: {
        type: Boolean,
        default: false,
      },
      whatsappEnabled: {
        type: Boolean,
        default: true,
      },
      emailEnabled: {
        type: Boolean,
        default: false,
      },
      notifyCustomerOnProcess: {
        type: Boolean,
        default: true,
      },
      notifyCustomerOnDelivery: {
        type: Boolean,
        default: true,
      },
      allowNegativeBalance: {
        type: Boolean,
        default: false,
      },
      lowBalanceThreshold: {
        type: Number,
        default: 100,
      },
    },
    lastLoginAt: {
      type: Date,
    },

    // Password support (added during week‑1 upgrade)
    passwordHash: {
      type: String,
    },

    // track login events (used for analytics/security)
    loginHistory: [
      {
        ip: String,
        userAgent: String,
        at: {
          type: Date,
          default: Date.now,
        },
      },
    ],

    truecallerId: {
      type: String,
      unique: true,
      sparse: true,
    },

    name: {
      type: String,
      trim: true,
    },
    role: {
      type: String,
      enum: ["admin", "vendor", "customer", "delivery_staff"],
      default: "vendor",
    },
  },
  {
    timestamps: true,
  }
);

const User = mongoose.model("User", userSchema);

export default User;

File: client/lib/features/auth/models/user_model.dart
class UserModel {
  const UserModel({
    required this.id,
    required this.phone,
    required this.role,
    this.name = '',
    this.businessName = '',
    this.ownerName = '',
    this.email = '',
    this.logoUrl = '',
    this.city = '',
    this.address = '',
    this.fcmToken = '',
    this.isActive = true,
  });

  final String id;
  final String phone;
  final String role;
  final String name;
  final String businessName;
  final String ownerName;
  final String email;
  final String logoUrl;
  final String city;
  final String address;
  final String fcmToken;
  final bool isActive;

  /// Vendor onboarding is complete when business + owner name exist (matches server onboarding).
  bool get isVendorProfileComplete =>
      businessName.trim().isNotEmpty && ownerName.trim().isNotEmpty;

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      phone: json['phone']?.toString() ?? '',
      role: json['role']?.toString() ?? 'vendor',
      name: json['name']?.toString() ?? '',
      businessName: json['businessName']?.toString() ?? '',
      ownerName: json['ownerName']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
      logoUrl: json['logoUrl']?.toString() ?? '',
      city: json['city']?.toString() ?? '',
      address: json['address']?.toString() ?? '',
      fcmToken: json['fcmToken']?.toString() ?? '',
      isActive: json['isActive'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (name.isNotEmpty) 'name': name,
      if (businessName.isNotEmpty) 'businessName': businessName,
      if (ownerName.isNotEmpty) 'ownerName': ownerName,
      if (email.isNotEmpty) 'email': email,
      if (logoUrl.isNotEmpty) 'logoUrl': logoUrl,
      if (city.isNotEmpty) 'city': city,
      if (address.isNotEmpty) 'address': address,
      'isActive': isActive,
      if (fcmToken.isNotEmpty) 'fcmToken': fcmToken,
    };
  }
}
